src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs

```diff
@@ -265,46 +293,130 @@ private BoundExpression MakePatternIndexOffsetExpression(BoundExpression unlower
             var lengthLocal = F.StoreToTemp(F.Property(receiverLocal, lengthOrCountProperty), out var lengthStore);
-            var rangeLocal = F.StoreToTemp(VisitExpression(rangeArg), out var rangeStore);
-            var startLocal = F.StoreToTemp(
-                F.Call(
-                    F.Call(rangeLocal, F.WellKnownMethod(WellKnownMember.System_Range__get_Start)),
-                    F.WellKnownMethod(WellKnownMember.System_Index__GetOffset),
-                    lengthLocal),
-                out var startStore);
-            var endLocal = F.StoreToTemp(
-                F.Call(
-                    F.Call(rangeLocal, F.WellKnownMethod(WellKnownMember.System_Range__get_End)),
-                    F.WellKnownMethod(WellKnownMember.System_Index__GetOffset),
-                    lengthLocal),
-                out var endStore);
+
+            localsBuilder.Add(receiverLocal.LocalSymbol);
+            sideEffectsBuilder.Add(receiverStore);
+
+            BoundExpression startExpr;
+            BoundExpression rangeSizeExpr;
+            if (rangeArg is BoundRangeExpression rangeExpr)
+            {
+                // If we know that the input is a range expression, we can
+                // optimize by pulling it apart inline, so
+                // 
+                // Range range = argumentExpr;
+                // int start = range.Start.GetOffset(length)
+                // int rangeSize = range.End.GetOffset(length) - start
+                //
+                // is, with `start..end`:
+                //
+                // int start = start.GetOffset(length)
+                // int rangeSize = end.GetOffset(length) - start
+
+                bool usedLength = false;
+
+                if (rangeExpr.LeftOperandOpt is null)
+                {
+                    startExpr = F.Literal(0);
+                }
+                else
+                {
+                    var startLocal = F.StoreToTemp(
+                        MakePatternIndexOffsetExpression(rangeExpr.LeftOperandOpt, lengthLocal, out usedLength),
+                        out var startStore);
+
+                    localsBuilder.Add(startLocal.LocalSymbol);
+                    sideEffectsBuilder.Add(startStore);
+                    startExpr = startLocal;
+                }
+
+                BoundExpression endExpr;
+                if (rangeExpr.RightOperandOpt is BoundExpression right)
+                {
+                    endExpr = MakePatternIndexOffsetExpression(
+                        right,
+                        lengthLocal,
+                        out bool usedLengthTemp);
+                    usedLength |= usedLengthTemp;
+                }
+                else
+                {
+                    usedLength = true;
+                    endExpr = lengthLocal;
+                }
+
+                if (usedLength)
+                {
+                    localsBuilder.Insert(1, lengthLocal.LocalSymbol);
```

Comment:
>Insert(1,  [](start = 34, length = 10)

why not `.Add(`? 
(also on next line)

Reply:
I believe side effects for the synthesized nodes are evaluated in order, so if we need the length, we need it *before* the bound nodes that use it, but *after* receiver, which is the first side effect calculated.